framework/src/auxkernels/NodalPatchRecovery.C

```diff
@@ -0,0 +1,224 @@
+  if (isNodal())
+  {
+    MeshBase & meshhelper = _mesh.getMesh();
+    meshhelper.allow_renumbering(false);
+    for (const auto & elem :
+         as_range(meshhelper.semilocal_elements_begin(), meshhelper.semilocal_elements_end()))
+      _fe_problem.addGhostedElem(elem->id());
+  }
+}
+
+std::vector<std::vector<unsigned>>
+NodalPatchRecovery::nchoosek(unsigned N, unsigned K)
+{
+  std::vector<std::vector<unsigned>> n_choose_k;
+  std::vector<unsigned> row;
+  std::string bitmask(K, 1); // K leading 1's
+  bitmask.resize(N, 0);      // N-K trailing 0's
+
+  do
+  {
+    row.clear();
+    row.push_back(0);
+    for (unsigned i = 0; i < N; ++i) // [0..N-1] integers
+      if (bitmask[i])
+        row.push_back(i + 1);
+    row.push_back(N + 1);
+    n_choose_k.push_back(row);
+  } while (std::prev_permutation(bitmask.begin(), bitmask.end()));
+
+  return n_choose_k;
+}
+
+std::vector<std::vector<unsigned>>
+NodalPatchRecovery::multiIndex(unsigned dim, unsigned order)
+{
+  // first row all zero
+  std::vector<std::vector<unsigned>> multi_index;
+  std::vector<std::vector<unsigned>> n_choose_k;
+  std::vector<unsigned> row(dim, 0);
+  multi_index.push_back(row);
+
+  if (dim == 1)
+    for (unsigned q = 1; q <= order; q++)
+    {
+      row[0] = q;
+      multi_index.push_back(row);
+    }
+  else
+    for (unsigned q = 1; q <= order; q++)
+    {
+      n_choose_k = nchoosek(dim + q - 1, dim - 1);
+      for (unsigned r = 0; r < n_choose_k.size(); r++)
+      {
+        row.clear();
+        for (unsigned c = 1; c < n_choose_k[0].size(); c++)
+          row.push_back(n_choose_k[r][c] - n_choose_k[r][c - 1] - 1);
+        multi_index.push_back(row);
+      }
+    }
+
+  return multi_index;
+}
+
+void
+NodalPatchRecovery::computePVector(Point q_point)
+{
+  _P.resize(_multi_index.size());
+  Real polynomial;
+  for (unsigned r = 0; r < _multi_index.size(); r++)
+  {
+    polynomial = 1.0;
+    for (unsigned c = 0; c < _multi_index[0].size(); c++)
+      for (unsigned p = 0; p < _multi_index[r][c]; p++)
+        polynomial *= q_point(c);
+    _P(r) = polynomial;
+  }
+}
+
+void
+NodalPatchRecovery::accumulateAMatrix()
+{
+  DenseMatrix<Number> PxP;
+  PxP.resize(_multi_index.size(), _multi_index.size());
+  PxP.outer_product(_P, _P);
+  _A.add(1.0, PxP);
+}
+
+void
+NodalPatchRecovery::accumulateBVector(Real val)
+{
+  _B.add(val, _P);
+}
+
+void
+NodalPatchRecovery::reinitPatch()
+{
+  // clean and resize P, A, B, coef
+  _P.resize(_multi_index.size());
+  _A.resize(_multi_index.size(), _multi_index.size());
+  _B.resize(_multi_index.size());
+  _coef.resize(_multi_index.size());
+
+  // activate dependent material properties
+  std::set<unsigned int> needed_mat_props;
+  const std::set<unsigned int> & mp_deps = getMatPropDependencies();
+  needed_mat_props.insert(mp_deps.begin(), mp_deps.end());
+  _fe_problem.setActiveMaterialProperties(needed_mat_props, _tid);
+}
+
+void
+NodalPatchRecovery::compute()
+{
+  if (!isNodal())
```

Comment:
A comment here indicating what we are doing with this switch would be helpful (just something simple like "Fall back on standard procedure for element variables", and "Use Zienkiewicz Zu patch recovery for nodal variables").